monitoring_service/token_network_listener.py

```diff
@@ -0,0 +1,119 @@
+        self.poll_interval = poll_interval
+        self.token_networks: Set[Address] = set()
+        self.token_network_listeners: List[BlockchainListener] = []
+        self.confirmed_channel_event_listeners: List[Callable] = []
+
+        log.info('Starting TokenNetworkRegistry Listener (required confirmations: {})...'.format(
+            self.required_confirmations,
+        ))
+        self.token_network_registry_listener = BlockchainListener(
+            web3=self.web3,
+            contract_manager=self.contract_manager,
+            contract_name=CONTRACT_TOKEN_NETWORK_REGISTRY,
+            contract_address=self.registry_address,
+            required_confirmations=self.required_confirmations,
+            poll_interval=self.poll_interval,
+            sync_start_block=self.sync_start_block,
+        )
+        log.info(
+            f'Listening to token network registry @ {registry_address} '
+            f'from block {sync_start_block}',
+        )
+        self.token_network_registry_listener.add_confirmed_listener(
+            topics=create_registry_event_topics(self.contract_manager),
+            callback=self.handle_token_network_created,
+        )
+
+    def handle_token_network_created(self, event: Dict):
+        token_network_address = event['args']['token_network_address']
+        token_address = event['args']['token_address']
+        event_block_number = event['blockNumber']
+
+        assert is_checksum_address(token_network_address)
+        assert is_checksum_address(token_address)
+
+        if token_network_address not in self.token_networks:
+            log.info(f'Found token network for token {token_address} @ {token_network_address}')
+
+            log.info('Creating token network for %s', token_network_address)
+            token_network_listener = BlockchainMonitor(
+                web3=self.web3,
+                contract_manager=self.contract_manager,
+                contract_address=token_network_address,
+                contract_name=CONTRACT_TOKEN_NETWORK,
+                required_confirmations=self.required_confirmations,
+                poll_interval=self.poll_interval,
+                sync_start_block=event_block_number,
+            )
+            token_network_listener.add_confirmed_listener(
+                topics=create_channel_event_topics(),
+                callback=self.make_handler(self.confirmed_channel_event_listeners),
+            )
+
+            token_network_listener.start()
+            self.token_networks.add(token_network_address)
+            self.token_network_listeners.append(token_network_listener)
+
+    def _run(self):
+        self.token_network_registry_listener.start()
+        self.stop_event.wait()
+
+    def stop(self):
+        self.token_network_registry_listener.stop()
+        for task in self.token_network_listeners:
+            task.stop()
+        self.stop_event.set()
+
+    def make_handler(self, listeners: List[Callable]):
```

Comment:
This seems a bit complicated, is there a reason to no do something like:
```py
def handle_event(self):
            for l in self.confirmed_channel_event_listeners:
                l(event, tx)
```
and set this as the callback?

Reply:
I had this for both confirmed and unconfirmed cases, in which case it removed the need to have two nearly identical handlers. Then I noticed that the `BlockchainMonitor` does not have a case for unconfirmed and removed that part for now.
We can go with the simple implementation for now.

Reply:
Makes sense now :)